### PR TITLE
First pass at making compile.js into TS

### PR DESCRIPTION
**Pull request by mattgodbolt**
This is a minimal PR to move some things (that may overlap with @dkm so I didn't go too deep) in one of the api handlers.

### CODE REVIEW

File: lib/handlers/compile.interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// IF YOU MODIFY ANYTHING HERE PLEASE UPDATE THE DOCUMENTATION!
+
+// This type models a request so all fields must be optional strings.
+export type CompileRequestQueryArgs = {
```
Comment (jeremy-rifkin) [inline lib/handlers/compile.interfaces.ts:28]: If we throw these types in the corresponding locations requests are created on the frontend, like `CompilerService.submit`, will they "just work"? 🤞

File: lib/base-compiler.ts [hunk 2]
```diff
@@ -2644,6 +2642,10 @@ but nothing was dumped. Possible causes are:
         }
     }
 
+    getModificationTime(): number | undefined {
+        return this.mtime ? this.mtime.getTime() : undefined;
```
[the remark this answers is not in the document]
Comment (mattgodbolt) [reply lib/base-compiler.ts:2646]: That doesn't compile (for reasons we don't understand still - maybe we need to change the language target, but running it causes errors)

File: lib/handlers/compile.interfaces.ts (hunk 1 above)
Comment (mattgodbolt) [reply lib/handlers/compile.interfaces.ts:28]: Maybe? I'm trying to limit the change as much as possible here for now

**Comment by partouf** [comment]
lgtm really

File: lib/base-compiler.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremy-rifkin) [reply lib/base-compiler.ts:2646]: Oh right

**Pull request merged**